addons/controllers/packageinstallstatus_controller.go

```diff
@@ -0,0 +1,442 @@
+	if tkr == nil {
+		log.Info("TKR object not found", "name", tkrName)
+		return ctrl.Result{}, nil
+	}
+
+	// determine cluster role
+	_, isManagementCluster := clusterLabels[tkrconstants.ManagementClusterRoleLabel]
+
+	if isManagementCluster {
+		// the cluster is a management cluster
+		clusterRole = clusterRoleManagement
+		c = r.Client
+	} else {
+		// the cluster is a remote workload cluster
+		clusterRole = clusterRoleWorkload
+		remoteClient, err := r.Tracker.GetClient(ctx, clusterapiutil.ObjectKey(cluster))
+		if err != nil {
+			log.Error(err, "error getting remote cluster's client")
+			return ctrl.Result{}, err
+		}
+		c = remoteClient
+		log.Info("successfully got remoteClient")
+		// set watch if not already set. If the watch already exists, it doesn't get re-created (logic in util.RemoteObjectTracker)
+		if err := r.watchPackageInstalls(ctx, cluster, log); err != nil {
+			log.Error(err, "error watching PackageInstalls on target cluster")
+			return ctrl.Result{}, err
+		}
+	}
+
+	if err := r.reconcile(ctx, c, cluster, clusterRole, log); err != nil {
+		return ctrl.Result{}, err
+	}
+
+	return ctrl.Result{}, nil
+}
+
+// reconcile iterates over all core/additional packages and calls reconcileClusterBootstrapStatus() for each.
+// it eventually updates ClusterBootstrapStatus with the condition entries (one for each package) in a single update operation
+// Note: use of the client c (remote or not) depends on whether the corresponding cluster is workload or management which is
+// determined in the Reconcile function
+func (r *PackageInstallStatusReconciler) reconcile(ctx context.Context, c client.Client, cluster *clusterapiv1beta1.Cluster, clusterRole ClusterRole, log logr.Logger) error {
+	clusterBootstrap := &runtanzuv1alpha3.ClusterBootstrap{}
+	// ClusterBootstrap resource exists in the management cluster, that's why we need to use local client instead of remote client for fetching it
+	if err := r.Client.Get(ctx, client.ObjectKeyFromObject(cluster), clusterBootstrap); err != nil {
+		return err
+	}
+	clusterBootstrap = clusterBootstrap.DeepCopy()
+
+	// this shouldn't include kapp ctrl package as it'll get processed separately
+	packages := append([]*runtanzuv1alpha3.ClusterBootstrapPackage{
+		clusterBootstrap.Spec.CNI,
+		clusterBootstrap.Spec.CPI,
+		clusterBootstrap.Spec.CSI,
+	}, clusterBootstrap.Spec.AdditionalPackages...)
+
+	var statusChanged bool
+
+	for _, pkg := range packages {
+		if pkg == nil {
+			continue
+		}
+		conditionChanged, err := r.reconcileClusterBootstrapStatus(ctx, c, &clusterBootstrap.Status, cluster.Name, clusterRole, pkg.RefName, constants.TKGSystemNS, log)
+		if err != nil {
+			// Note: just log the error and continue with collecting PackageInstallStatus for other packages
+			log.Error(err, fmt.Sprintf("failed to reconcile PackageInstallStatus for package '%s/%s'", constants.TKGSystemNS, pkg.RefName))
+		}
+		statusChanged = statusChanged || conditionChanged
```

Comment:
Nit: The naming for `conditionChanged` is a bit confusing here. I would suggest something like `packageStatusChanged`.

Reply:
I think conditionChanged is more correct, as it refers to an entry in the ClusterBootstrap conditions not PackageInstall status

Reply:
Changed to ```cbConditionChanged``` and ```cbStatusChanged```